app/Services/RunwayService.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+namespace App\Services;
+
+use App\Exceptions\Airfield\AirfieldNotFoundException;
+use App\Exceptions\Runway\RunwayHeadingInvalidException;
+use App\Exceptions\Runway\RunwayIdentifierInvalidException;
+use App\Exceptions\Runway\RunwayThresholdInvalidException;
+use App\Helpers\Sectorfile\Coordinate;
+use App\Models\Airfield\Airfield;
+use App\Models\Runway\Runway;
+use App\Rules\Runway\RunwayIdentifier;
+use Exception;
+use Illuminate\Support\Facades\DB;
+
+class RunwayService
+{
+    public static function addRunwayPair(
+        string $airfield,
+        string $firstIdentifier,
+        int $firstHeading,
+        Coordinate $firstThreshold,
+        string $secondIdentifier,
+        int $secondHeading,
+        Coordinate $secondThreshold
+    ): void {
+        $airfieldModel = Airfield::where('code', $airfield)
+            ->first();
+        if (!$airfieldModel) {
+            throw AirfieldNotFoundException::fromIcao($airfield);
+        }
+
+        self::validateRunway($firstIdentifier, $firstHeading, $firstThreshold);
+        self::validateRunway($secondIdentifier, $secondHeading, $secondThreshold);
+
+        if (!self::headingsMatch($firstHeading, $secondHeading)) {
+            throw RunwayHeadingInvalidException::forHeadings($firstHeading, $secondHeading);
+        }
+
+        // Create the runways, then link them as "inverse" runways
+        DB::transaction(
```

Comment:
Are you sure that inside of this will roll back if it fails? Does it not rely on all calls being through the DB facade?

Reply:
From the documentation :)

```
The DB facade's transaction methods control the transactions for both the query builder and Eloquent ORM.
```